controller-server/src/main/java/com/yahoo/vespa/hosted/controller/ApplicationController.java

```diff
@@ -505,31 +505,42 @@ private void registerRotationInDns(Rotation rotation, String dnsName) {
     }
 
     /**
-     * Deletes the application with this id
+     * Deletes the the given application. All known instances of the applications will be deleted,
+     * including PR instances.
      * 
      * @throws IllegalArgumentException if the application has deployments or the caller is not authorized
-     * @throws NotExistsException if the application does not exist
+     * @throws NotExistsException if no instances of the application exist
      */
-    public void deleteApplication(ApplicationId id, Optional<NToken> token) {
-        if ( ! controller.applications().get(id).isPresent())
-            throw new NotExistsException("Could not delete application '" + id + "': Application not found");
+    public void deleteApplication(ApplicationId applicationId, Optional<NToken> token) {
+        // Find all instances of the application
+        List<ApplicationId> instances = controller.applications().asList(applicationId.tenant())
+                                                  .stream()
+                                                  .map(Application::id)
+                                                  .filter(id -> id.application().equals(applicationId.application()) &&
+                                                                id.tenant().equals(applicationId.tenant()))
+                                                  .collect(Collectors.toList());
+        if (instances.isEmpty()) {
+            throw new NotExistsException("Could not delete application '" + applicationId + "': Application not found");
+        }
 
-        lockOrThrow(id, application -> {
+        // TODO: Make this one transaction when database is moved to ZooKeeper
+        instances.forEach(id -> lockOrThrow(id, application -> {
             if ( ! application.deployments().isEmpty())
                 throw new IllegalArgumentException("Could not delete '" + application + "': It has active deployments");
-            
+
             Tenant tenant = controller.tenants().tenant(new TenantId(id.tenant().value())).get();
             if (tenant.isAthensTenant() && ! token.isPresent())
                 throw new IllegalArgumentException("Could not delete '" + application + "': No NToken provided");
 
-            // NB: Next 2 lines should have been one transaction
-            if (tenant.isAthensTenant())
+            if (tenant.isAthensTenant()) {
```

Comment:
I think this clause should be moved below the loop.

Reply:
Done. Was easier to do it in the loop, but only for the default instance.